### Bibliotecas necessárias

In [1]:
# realizando o download e instalação da biblioteca
! pip install nltk

In [52]:
# importando as bibliotecas necessárias
import nltk
from sklearn.feature_extraction.text import TfidfVectorizer
import pandas as pd

### Pacotes necessários

In [4]:
# importando os pacotes "stopwords" e "punkt"
nltk.download('stopwords')
nltk.download('punkt')

[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Unzipping corpora/stopwords.zip.
[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Unzipping tokenizers/punkt.zip.


True

Use a variável texto a seguir para identificar as stop words
e após isso aplique o cáculo do **TF-IDF** neste mesmo texto e observe os scores.

**OBS:** Para realizar essa tarefa utilize os módulos estudados na aula

In [66]:
# instaciando um texto na variável "texto"
texto = 'Um carro azul seguia rapidamente em uma rodovia, e ao passar por um buraco, o carro furou o pneu, e o motorista desceu do carro azul'

In [67]:
# instanciando o algoritmo na variável "stopwords"
# e configurando para o obte-las em português
stopwords = nltk.corpus.stopwords.words('portuguese')

In [68]:
# para conhecer as stopwords em português vamos imprimi-las
print(stopwords)

['de', 'a', 'o', 'que', 'e', 'é', 'do', 'da', 'em', 'um', 'para', 'com', 'não', 'uma', 'os', 'no', 'se', 'na', 'por', 'mais', 'as', 'dos', 'como', 'mas', 'ao', 'ele', 'das', 'à', 'seu', 'sua', 'ou', 'quando', 'muito', 'nos', 'já', 'eu', 'também', 'só', 'pelo', 'pela', 'até', 'isso', 'ela', 'entre', 'depois', 'sem', 'mesmo', 'aos', 'seus', 'quem', 'nas', 'me', 'esse', 'eles', 'você', 'essa', 'num', 'nem', 'suas', 'meu', 'às', 'minha', 'numa', 'pelos', 'elas', 'qual', 'nós', 'lhe', 'deles', 'essas', 'esses', 'pelas', 'este', 'dele', 'tu', 'te', 'vocês', 'vos', 'lhes', 'meus', 'minhas', 'teu', 'tua', 'teus', 'tuas', 'nosso', 'nossa', 'nossos', 'nossas', 'dela', 'delas', 'esta', 'estes', 'estas', 'aquele', 'aquela', 'aqueles', 'aquelas', 'isto', 'aquilo', 'estou', 'está', 'estamos', 'estão', 'estive', 'esteve', 'estivemos', 'estiveram', 'estava', 'estávamos', 'estavam', 'estivera', 'estivéramos', 'esteja', 'estejamos', 'estejam', 'estivesse', 'estivéssemos', 'estivessem', 'estiver', 'estiv

**Tokenização**
- vamos realizar a tokenização e remover as stopwords

In [69]:
# importando a biblioteca
from nltk.tokenize import word_tokenize

In [70]:
# realizando a divisão da frase criada em tokens
tokens = word_tokenize(texto)

# imprimindo a tokenização
print(tokens)

['Um', 'carro', 'azul', 'seguia', 'rapidamente', 'em', 'uma', 'rodovia', ',', 'e', 'ao', 'passar', 'por', 'um', 'buraco', ',', 'o', 'carro', 'furou', 'o', 'pneu', ',', 'e', 'o', 'motorista', 'desceu', 'do', 'carro', 'azul']


In [71]:
# realizando um loop simples para imprimir a frase sem os stopwords e melhorar sua visualização
for t in tokens:
  if t not in stopwords:
    print(t)

Um
carro
azul
seguia
rapidamente
rodovia
,
passar
buraco
,
carro
furou
pneu
,
motorista
desceu
carro
azul


Agora vamos utilizar a técnica de cálculo estatístico, `TF-IDF` (do inglês Term Frequency – Inverse Document Frequency) que é muito importante na análise de textos. Pois ela identifica a frequência e a importância que uma palavra para o texto analisado.


In [72]:
# importando a biblioteca
from sklearn.feature_extraction.text import TfidfVectorizer

In [73]:
# instanciando o módulo
tf_idf = TfidfVectorizer()

**Usando o fit_transform**

E instanciando numa variável, onde ele retornará uma matriz com os termos e os scores associados, passando o `texto` como contéudo.

In [76]:
# aplicando o ".fit_transform"
vetor = tf_idf.fit_transform([texto])

# imprimindo a variável
print(vetor) # retorna uma matriz com os termos e os scores associados

  (0, 5)	0.1796053020267749
  (0, 4)	0.1796053020267749
  (0, 8)	0.1796053020267749
  (0, 10)	0.1796053020267749
  (0, 7)	0.1796053020267749
  (0, 2)	0.1796053020267749
  (0, 11)	0.1796053020267749
  (0, 9)	0.1796053020267749
  (0, 0)	0.1796053020267749
  (0, 13)	0.1796053020267749
  (0, 16)	0.1796053020267749
  (0, 6)	0.1796053020267749
  (0, 12)	0.1796053020267749
  (0, 14)	0.1796053020267749
  (0, 1)	0.3592106040535498
  (0, 3)	0.5388159060803247
  (0, 15)	0.3592106040535498


In [77]:
# colocando a matriz em um formato de array
vetor = vetor.todense()

print(vetor)

[[0.1796053  0.3592106  0.1796053  0.53881591 0.1796053  0.1796053
  0.1796053  0.1796053  0.1796053  0.1796053  0.1796053  0.1796053
  0.1796053  0.1796053  0.1796053  0.3592106  0.1796053 ]]


In [64]:
# obtendo os nomes das palavras mapeadas
nomes  = tf_idf.get_feature_names()

# imprimindo a variável
print(nomes)

['ao', 'azul', 'buraco', 'carro', 'desceu', 'do', 'em', 'furou', 'motorista', 'passar', 'pneu', 'por', 'rapidamente', 'rodovia', 'seguia', 'um', 'uma']


In [25]:
# associando os scores aos nomes das palavras e o resultado em um dataframe
df = pd.DataFrame(vetor, columns=nomes)
print(df)

         ao      azul    buraco  ...    seguia        um       uma
0  0.179605  0.359211  0.179605  ...  0.179605  0.359211  0.179605

[1 rows x 17 columns]


In [49]:
df.max()

ao             0.179605
azul           0.359211
buraco         0.179605
carro          0.538816
desceu         0.179605
do             0.179605
em             0.179605
furou          0.179605
motorista      0.179605
passar         0.179605
pneu           0.179605
por            0.179605
rapidamente    0.179605
rodovia        0.179605
seguia         0.179605
um             0.359211
uma            0.179605
dtype: float64

**Perceba que, a palavra com maior importância é** `"carro"`**, com score de** `0.538816` **e, as seguintes palavras são** `"azul"` **e** `"um"` **com scores de** `0.359211`**.**

In [50]:
# df.values

In [51]:
# df.columns